fastcrypto-zkp/src/bn254/zk_login.rs

```diff
@@ -1,86 +1,144 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use ark_crypto_primitives::snark::SNARK;
-use std::collections::HashMap;
-use std::fmt;
-
-use super::{poseidon::PoseidonWrapper, verifier::process_vk_special};
-use crate::bn254::VerifyingKey as Bn254VerifyingKey;
-use crate::circom::CircomPublicInputs;
-use crate::{
-    bn254::verifier::PreparedVerifyingKey,
-    circom::{g1_affine_from_str_projective, g2_affine_from_str_projective},
-};
+use fastcrypto::error::FastCryptoResult;
+use serde_json::Value;
+
+use super::poseidon::PoseidonWrapper;
+use crate::circom::{g1_affine_from_str_projective, g2_affine_from_str_projective};
 pub use ark_bn254::{Bn254, Fr as Bn254Fr};
 pub use ark_ff::ToConstraintField;
-use ark_groth16::{Groth16, Proof, VerifyingKey};
+use ark_groth16::Proof;
 pub use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
 use fastcrypto::{
     error::FastCryptoError,
     rsa::{Base64UrlUnpadded, Encoding},
 };
-use num_bigint::{BigInt, BigUint};
-use once_cell::sync::Lazy;
-use regex::Regex;
+use num_bigint::{BigInt, BigUint, Sign};
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
-use serde_json::Value;
 use std::str::FromStr;
 
+type ParsedJWKs = Vec<((String, String), OAuthProviderContent)>;
 #[cfg(test)]
 #[path = "unit_tests/zk_login_tests.rs"]
 mod zk_login_tests;
 
-static GLOBAL_VERIFYING_KEY: Lazy<PreparedVerifyingKey> = Lazy::new(global_pvk);
-
-/// Hardcoded mapping from the provider and its supported key claim name to its map-to-field Big Int in string.
-/// The field value is computed from the max key claim length and its provider.
-static SUPPORTED_KEY_CLAIM_TO_FIELD: Lazy<HashMap<(&str, String), &str>> = Lazy::new(|| {
-    let mut map = HashMap::new();
-    map.insert(
-        (
-            OAuthProvider::Google.get_config().0,
-            SupportedKeyClaim::Sub.to_string(),
-        ),
-        "18523124550523841778801820019979000409432455608728354507022210389496924497355",
-    );
-    map.insert(
-        (
-            OAuthProvider::Google.get_config().0,
-            SupportedKeyClaim::Email.to_string(),
-        ),
-        "",
-    );
-    map.insert(
-        (
-            OAuthProvider::Twitch.get_config().0,
-            SupportedKeyClaim::Sub.to_string(),
-        ),
-        "",
-    );
-    map.insert(
-        (
-            OAuthProvider::Twitch.get_config().0,
-            SupportedKeyClaim::Email.to_string(),
-        ),
-        "",
-    );
-    map
-});
+const MAX_HEADER_LEN: u16 = 500;
+const PACK_WIDTH: u16 = 248;
+const ISS: &str = "iss";
+const AUD: &str = "aud";
+const NUM_EXTRACTABLE_STRINGS: u8 = 5;
+const MAX_EXTRACTABLE_STR_LEN: u16 = 150;
+const MAX_EXTRACTABLE_STR_LEN_B64: u16 = 4 * (1 + MAX_EXTRACTABLE_STR_LEN / 3);
 
 /// Supported OAuth providers. Must contain "openid" in "scopes_supported"
 /// and "public" for "subject_types_supported" instead of "pairwise".
```

Comment:
The statement about "public" vs "pairwise" is not true anymore. We support both.